### PR TITLE
validate dependent jobs' parents - issue 148

### LINKED ISSUE #148: we should validate that a given service.instance parent exists before trying to deploy
We currently only check for the 'parent' key of a job being useful when we attempt to launch the job. Instead, this should (also?) be done as part of a config's validation.

### PR DESCRIPTION
**Pull request by giuliano108**
This should fix #148 .

Catches two types of config errors:
- Parents whose name is invalid (!= `service_name.instance_name`). Similar to what the JSON schema also enforces.
- Parents that are referenced in a dependent job, but are not actually configured.

Note that the second validation won't let a "`ServiceA` child" depend on a "`ServiceB` parent". That's the best we can do without parsing the entire `yelpsoa-configs` tree.

In my opinion `paasta_tools/cli/cmds/validate.py`, in order to validate the Chronos config, needs to know too many details about `ChronosJobConfig`. If need be, I'm happy to move as much of that logic into `paasta_tools/chronos_tools.py` as possible, so that the latter is in charge of the validation and the former just prints out the outcome.

screen shot 2016-02-10 at 6 08 14 pm

### CODE REVIEW

File: paasta_tools/cli/cmds/validate.py [hunk 1]
```diff
@@ -205,16 +207,27 @@ def validate_chronos(service_path):
                 service=service, clusters=[cluster], instance_type=instance_type,
                 soa_dir=soa_dir):
             cjc = load_chronos_job_config(service, instance, cluster, False, soa_dir)
-            checks_passed, check_msgs = cjc.validate()
+            cjcs[service + '.' + cjc.get_job_name()] = cjc
```
Comment (solarkennedy) [inline paasta_tools/cli/cmds/validate.py]: This looks a little more complicated than I think it needs to be. While not super efficient, can it just be:
```
for parent in get_parents()
    if parent not in get_services_for_cluster(cluster=cluster, instance_type=chronos):
        # something
```

**Comment by giuliano108** [comment]
> This looks a little more complicated than I think it needs to be.

Yes. Using `get_services_for_cluster` also allows parent/dependent validation to work across services in the same cluster. In this new revision, I'm also making sure validation fails when a dependent job specifies itself as a parent.

screen shot 2016-02-15 at 6 42 53 pm

File: paasta_tools/cli/cmds/validate.py [hunk 2]
```diff
@@ -201,12 +203,25 @@ def validate_chronos(service_path):
 
     returncode = True
     for cluster in list_clusters(service, soa_dir, instance_type):
+        services_in_cluster = get_services_for_cluster(cluster=cluster, instance_type='chronos')
+        valid_services = set([name + '.' + instance for name, instance in services_in_cluster])
```
Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/validate.py]: "%s%s%s" % (service, ".", instance)

File: paasta_tools/cli/cmds/validate.py [hunk 3]
```diff
@@ -201,12 +203,25 @@ def validate_chronos(service_path):
...
                 service=service, clusters=[cluster], instance_type=instance_type,
                 soa_dir=soa_dir):
             cjc = load_chronos_job_config(service, instance, cluster, False, soa_dir)
+            parents = cjc.get_parents() or []
             checks_passed, check_msgs = cjc.validate()
 
+            for parent in parents:
+                if not check_parent_format(parent):
+                    continue
+                if service + '.' + instance == parent:
```
Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/validate.py]: "%s%s%s" % (service, ".", instance)

**Comment by solarkennedy** [comment]
lgtm

**Comment by Rob-Johnson** [comment]
fix n ship

**Pull request merged**